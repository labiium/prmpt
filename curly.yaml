base:
  path: .
  ignore:
    - '*.tmp'
    - 'temp'
    - test_dir
    - test
  # output: 'curly.out'
  delimiter: '```'
  language: rust
  use_gitignore: true
  # prompts:
  #   - 'Provide the relative file path at the top of the block.'
  #   - 'Follow the relative file path with the code, ensuring there is no additional text in between.'
  #   - 'Think logically, breaking down the problem step by step within the comments of the code.'
advanced:
  path: 'src'
  ignore:
    - '*.bak'
  output: 'advanced_out'
  prompts:
    - 'Use advanced terminology.'
