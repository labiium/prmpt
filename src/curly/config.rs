//! Holds the configuration structure (`Config`) and functionality to load configurations.

use serde::{Deserialize, Serialize};
use std::{collections::HashMap, fs, path::Path};

/// Configuration structure that holds various options for generating or injecting code.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Config {
    /// Path to the code repository.
    pub path: Option<String>,
    /// Glob patterns to ignore.
    pub ignore: Option<Vec<String>>,
    /// File path to write the generated prompt.
    pub output: Option<String>,
    /// Delimiter for code blocks in the prompt (e.g., "```").
    pub delimiter: Option<String>,
    /// The programming language of the repository (e.g. "rust", "python").
    pub language: Option<String>,
    /// Additional prompts that can be injected into the output for specific files.
    pub prompts: Option<Vec<String>>,
    /// If true, only documentation and comments are extracted (used for e.g. docs-only runs).
    pub docs_comments_only: Option<bool>,
    /// Patterns to ignore specifically in documentation comments.
    pub docs_ignore: Option<Vec<String>>,
    /// If true, respects patterns in a `.gitignore` file.
    pub use_gitignore: Option<bool>,
    /// If true, any outputs from Jupyter Notebook cells will be included in the generated prompt.
    pub display_outputs: Option<bool>,
}

pub const DEFAULT_CONFIG_KEY: &str = "base";

/// Loads configuration from a local `curly.yaml` file.
/// The file can contain a single configuration or multiple named configurations.
pub fn load_config() -> Result<HashMap<String, Config>, Box<dyn std::error::Error>> {
    let config_path = Path::new("curly.yaml");
    let contents = fs::read_to_string(config_path)?;

    // Parse the YAML generically first so we can determine its structure
    let yaml_value: serde_yaml::Value = serde_yaml::from_str(&contents)?;
    let mapping = yaml_value.as_mapping().ok_or("curly.yaml must contain a mapping at the top level")?;

    // Set of valid Config field names to distinguish between a single config and a map of configs
    const CONFIG_FIELDS: &[&str] = &[
        "path",
        "ignore",
        "output",
        "delimiter",
        "language",
        "prompts",
        "docs_comments_only",
        "docs_ignore",
        "use_gitignore",
        "display_outputs",
    ];

    let all_keys_are_fields = mapping.keys().all(|k| {
        k.as_str()
            .map(|key| CONFIG_FIELDS.contains(&key))
            .unwrap_or(false)
    });

    if all_keys_are_fields {
        // Deserialize directly into a single Config
        let single_config: Config = serde_yaml::from_value(yaml_value)?;
        let mut configs = HashMap::new();
        configs.insert(DEFAULT_CONFIG_KEY.to_string(), single_config);
        Ok(configs)
    } else {
        // Treat each top level key as a named configuration
        let multiple_configs: HashMap<String, Config> = serde_yaml::from_value(yaml_value)?;
        Ok(multiple_configs)
    }
}
