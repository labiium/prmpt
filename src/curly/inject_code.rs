//! Provides functionality for injecting code from a generated file (e.g., "curly.out")
//! back into the repository at specified file paths.

use log::{error, info, warn};
use std::{
    fs, // Removed io
    path::Path,
};
use rand::Rng;
use rand::distr::Alphanumeric; 
use rand::rngs::ThreadRng;
use crate::curly::traits::InjectOperation; // Import the trait
use anyhow::{Context, Error}; // For the Result type & context

/// Struct for implementing the InjectOperation trait.
#[derive(Default)]
pub struct Injector;

impl InjectOperation for Injector {
    /// Injects code from a specified input file into a target repository path.
    /// This method encapsulates the original `inject` function's logic.
    fn inject(&self, input_path: &Path, repo_path: &Path) -> Result<(), Error> {
        // Canonicalize the base repo path
        let base_path_canon = fs::canonicalize(repo_path)
            .with_context(|| format!("Failed to canonicalize base repository path: '{}'", repo_path.display()))?;
        info!("Canonicalized base repository path: {:?}", base_path_canon);

        let contents = fs::read_to_string(input_path)
            .with_context(|| format!("Failed to read input file: '{}'", input_path.display()))?;
        
        let delimiter = "```";
        let mut lines = contents.lines();
        let mut current_file_target_path_str: Option<String> = None;
        let mut code_block = String::new();
        let mut in_code_block = false;

        info!("Starting to process the input file for injection: {:?}", input_path);

        while let Some(line) = lines.next() {
            if !in_code_block
                && ((line.trim_start().starts_with("### `") && line.trim_end().ends_with('`'))
                    || (line.trim_start().starts_with("**`") && line.trim_end().ends_with("`**"))
                    || (line.trim_start().starts_with('`')
                        && line.trim_end().ends_with('`')
                        && line.len() > 3))
            {
                let extracted_path_str = extract_path(line);
                if !extracted_path_str.trim().is_empty() {
                    current_file_target_path_str = Some(extracted_path_str.to_string());
                    info!("Detected relative file path for injection: {:?}", current_file_target_path_str);
                } else {
                    warn!("Detected an empty file path! Skipping...");
                    current_file_target_path_str = None;
                }
            }
            else if line.trim_start().starts_with(delimiter) {
                // Handle optional path on the same line as the opening code fence
                if !in_code_block {
                    if let Some(path_on_fence) = extract_path_from_fence(line, delimiter) {
                        current_file_target_path_str = Some(path_on_fence.to_string());
                        info!("Detected relative file path for injection: {:?}", current_file_target_path_str);
                    }
                }
                in_code_block = !in_code_block;
                if !in_code_block { // Closing a code block
                    if let Some(ref target_file_rel_str) = current_file_target_path_str {
                        if code_block.is_empty() {
                            warn!("Empty code block detected for path: {:?}", target_file_rel_str);
                            current_file_target_path_str = None; 
                            continue;
                        }

                        let full_target_path = base_path_canon.join(target_file_rel_str);
                        let target_filename = match full_target_path.file_name() {
                            Some(name) => name.to_os_string(),
                            None => {
                                error!("Could not extract filename from path: {:?}", full_target_path);
                                current_file_target_path_str = None; code_block.clear(); continue;
                            }
                        };
                        
                        let parent_dir_for_file = full_target_path.parent().unwrap_or_else(|| Path::new(""));

                        // Ensure parent directory exists, trying to canonicalize it.
                        let canonical_parent_dir = if parent_dir_for_file.as_os_str().is_empty() || parent_dir_for_file == base_path_canon.as_path() {
                             base_path_canon.clone()
                        } else if !parent_dir_for_file.is_absolute() && parent_dir_for_file.starts_with(&base_path_canon) {
                            // If parent_dir_for_file is already relative to base_path_canon correctly
                            fs::create_dir_all(&parent_dir_for_file)
                                .with_context(|| format!("Failed to create parent directory: {:?}", parent_dir_for_file))?;
                            fs::canonicalize(&parent_dir_for_file)
                                .with_context(|| format!("Failed to canonicalize parent directory: {:?}", parent_dir_for_file))?
                        } else {
                             // This case handles when full_target_path.parent() is absolute or needs careful joining
                             // For simplicity, assume it's relative or directly under base_path_canon as handled by base_path_canon.join()
                             // If it's already absolute and within repo_path, fs::create_dir_all should handle it.
                             // This part might need more robust handling for complex symlink scenarios outside typical usage.
                             fs::create_dir_all(parent_dir_for_file)
                                 .with_context(|| format!("Failed to create parent directory: {:?}", parent_dir_for_file))?;
                             fs::canonicalize(parent_dir_for_file)
                                 .with_context(|| format!("Failed to canonicalize parent directory: {:?}", parent_dir_for_file))?
                        };


                        let final_file_path_canon = canonical_parent_dir.join(&target_filename);
                        info!("Final canonical file path for injection: {:?}", final_file_path_canon);

                        let mut rng = ThreadRng::default();
                        let random_string: String = (&mut rng)
                            .sample_iter(&Alphanumeric)
                            .take(6)
                            .map(char::from)
                            .collect();
                        let temp_filename = format!(".{}.tmp.{}", target_filename.to_string_lossy(), random_string);
                        let temp_file_path = canonical_parent_dir.join(temp_filename);

                        info!("Writing to temporary file: {:?}", temp_file_path);
                        match fs::write(&temp_file_path, code_block.trim_end()) {
                            Ok(_) => {
                                info!("Successfully wrote to temporary file. Renaming to: {:?}", final_file_path_canon);
                                if let Err(e) = fs::rename(&temp_file_path, &final_file_path_canon) {
                                    let rename_err = Error::new(e).context(format!("Failed to rename temporary file {:?} to {:?}", temp_file_path, final_file_path_canon));
                                    error!("{:?}", rename_err);
                                    if let Err(remove_err) = fs::remove_file(&temp_file_path) {
                                        error!("Additionally, failed to remove temporary file {:?}: {}", temp_file_path, remove_err);
                                    }
                                    // Decide if this is a critical error for the whole inject operation
                                    // For now, log and continue, but one could return rename_err here.
                                } else {
                                    info!("Successfully injected code into {:?}", final_file_path_canon);
                                }
                            }
                            Err(e) => {
                                error!("Failed to write to temporary file {:?}: {}", temp_file_path, e);
                                if temp_file_path.exists() {
                                    if let Err(remove_err) = fs::remove_file(&temp_file_path) {
                                        error!("Additionally, failed to remove temporary file {:?}: {}", temp_file_path, remove_err);
                                    }
                                }
                            }
                        }
                        code_block.clear();
                    } else {
                        warn!("Code block closed without a file path being set!");
                    }
                    current_file_target_path_str = None; 
                } else {
                    info!("Entering a code block...");
                    code_block.clear();
                }
            }
            else if !in_code_block {
                let trimmed = line.trim();
                if !trimmed.is_empty()
                    && !trimmed.starts_with('#')
                    && !trimmed.starts_with(delimiter)
                    && !trimmed.contains(' ')
                {
                    current_file_target_path_str = Some(trimmed.to_string());
                    info!("Detected relative file path for injection: {:?}", current_file_target_path_str);
                }
            }
            else if in_code_block {
                code_block.push_str(line);
                code_block.push('\n');
            }
        }
        info!("Finished processing the input file for injection.");
        Ok(())
    }
}

// The old `inject` function is removed as its logic is now in `Injector::inject`.

/// Helper function for extracting the path from a line
/// that looks like `### `path/to/file` or **`path/to/file`**, etc.
fn extract_path(input: &str) -> &str {
    // Trim leading/trailing whitespace which might affect path extraction
    let trimmed_input = input.trim();
    if trimmed_input.starts_with("### `") && trimmed_input.ends_with('`') {
        &trimmed_input[5..trimmed_input.len() - 1]
    } else if trimmed_input.starts_with("**`") && trimmed_input.ends_with("`**") {
        &trimmed_input[3..trimmed_input.len() - 3]
    } else if trimmed_input.starts_with('`') && trimmed_input.ends_with('`') { // Generic backtick case
        &trimmed_input[1..trimmed_input.len() - 1]
    } else {
        trimmed_input // Fallback if no known pattern matches, assume the line itself is the path
    }
} // Added missing closing brace

/// Attempt to extract a file path from a line that begins with the code block
/// delimiter. This supports Curly's own output format where the file path
/// directly follows the opening fence, e.g. "```src/lib.rs" or
/// "```rust src/lib.rs".
fn extract_path_from_fence<'a>(line: &'a str, delimiter: &str) -> Option<&'a str> {
    let remainder = line.trim_start().strip_prefix(delimiter)?.trim();
    if remainder.is_empty() {
        return None;
    }
    // If multiple tokens exist after the delimiter, assume the last one is the path
    let tokens: Vec<&str> = remainder.split_whitespace().collect();
    if tokens.len() == 1 {
        let t = tokens[0];
        if t.contains('/') || t.contains('.') {
            return Some(t);
        }
        return None;
    }
    tokens.last().copied()
}
