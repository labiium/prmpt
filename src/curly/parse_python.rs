//! Implements functionality for parsing Python code using Tree-sitter in order
//! to extract function/class signatures and docstrings.

use serde_json::Value as JsonValue;
use std::fs;
use tree_sitter::{Node, Parser};
use tree_sitter_python;

/// Sets up a Tree-sitter parser for Python and extracts function/class signatures
/// along with docstrings from the provided file contents.
///
/// This is used in "docs-only" modes, or to produce more descriptive prompts.
pub fn extract_python_signatures(contents: &str) -> String {
    let mut parser = Parser::new();
    let language = tree_sitter_python::language();
    parser
        .set_language(language)
        .expect("Error loading Python grammar");

    let tree = parser.parse(contents, None).unwrap();
    let root_node = tree.root_node();

    let mut signatures = String::new();

    // Before processing other nodes, check for module-level docstring
    let module_docstring = extract_module_docstring(root_node, contents);
    if !module_docstring.is_empty() {
        signatures.push_str(&module_docstring);
        signatures.push('\n');
    }

    // Process definitions starting from the root node
    let mut cursor = root_node.walk();
    for child in root_node.children(&mut cursor) {
        let child_signatures = extract_definitions(child, contents, 0);
        if !child_signatures.is_empty() {
            signatures.push_str(&child_signatures);
            signatures.push('\n');
        }
    }

    signatures
}

/// Attempts to extract a module-level docstring from a Python file.
fn extract_module_docstring(root_node: Node, source_code: &str) -> String {
    let mut cursor = root_node.walk();
    for child in root_node.children(&mut cursor) {
        if child.kind() == "expression_statement" {
            let mut expr_cursor = child.walk();
            for expr_child in child.named_children(&mut expr_cursor) {
                if expr_child.kind() == "string" {
                    let docstring_text = expr_child.utf8_text(source_code.as_bytes()).unwrap();
                    let (stripped_docstring, quote_type) = strip_quotes(docstring_text);
                    let indented_docstring = indent_docstring(stripped_docstring, "", quote_type);
                    return indented_docstring;
                }
            }
        } else if !child.kind().starts_with("comment") && !child.kind().starts_with("newline") {
            // If we encounter anything else, no module docstring is present
            break;
        }
    }
    String::new()
}

/// Recursively extracts function and class signatures from a given AST node.
fn extract_definitions(node: Node, source_code: &str, indent_level: usize) -> String {
    let mut output = String::new();

    if node.kind() == "function_definition" || node.kind() == "class_definition" {
        let signature = extract_signature_and_docstring(node, source_code, indent_level);
        output.push_str(&signature);
        output.push('\n');
    }

    // If this is a class or function, process its body to find nested definitions
    if node.kind() == "class_definition" || node.kind() == "function_definition" {
        // Find the block node (the body)
        let mut cursor = node.walk();
        for child in node.children(&mut cursor) {
            if child.kind() == "block" {
                // Process the statements inside the block
                let mut block_cursor = child.walk();
                for stmt in child.children(&mut block_cursor) {
                    let stmt_output = extract_definitions(stmt, source_code, indent_level + 1);
                    if !stmt_output.is_empty() {
                        output.push_str(&stmt_output);
                        output.push('\n');
                    }
                }
            }
        }
    }

    output
}

/// Extracts the signature (including decorators) and docstring from a function or class node.
fn extract_signature_and_docstring(node: Node, source_code: &str, indent_level: usize) -> String {
    let mut signature = String::new();
    let indent = "    ".repeat(indent_level);

    if node.kind() == "function_definition" || node.kind() == "class_definition" {
        signature.push_str(&indent);

        let mut cursor = node.walk();
        let children = node.children(&mut cursor);
        let mut found_signature = false;
        for child in children {
            match child.kind() {
                "decorator" => {
                    let decorator_text = child.utf8_text(source_code.as_bytes()).unwrap();
                    signature.push_str(&indent);
                    signature.push_str(decorator_text);
                    signature.push('\n');
                }
                "def" | "class" => {
                    let text = child.utf8_text(source_code.as_bytes()).unwrap();
                    signature.push_str(text);
                    signature.push(' ');
                }
                "identifier" | "parameters" | ":" => {
                    let text = child.utf8_text(source_code.as_bytes()).unwrap();
                    signature.push_str(text);
                    if child.kind() == ":" {
                        signature.push('\n');
                    }
                    found_signature = true;
                }
                "block" => {
                    // After the signature, look for the docstring
                    if found_signature {
                        let docstring = extract_docstring(child, source_code, indent_level + 1);
                        if !docstring.is_empty() {
                            signature.push_str(&docstring);
                        }
                        break; // We only need the first block
                    }
                }
                _ => {}
            }
        }
    }

    signature
}

/// Attempts to extract a docstring from a code block (e.g., the first statement in a function).
fn extract_docstring(block_node: Node, source_code: &str, indent_level: usize) -> String {
    let mut cursor = block_node.walk();
    let mut children = block_node.named_children(&mut cursor);

    let indent = "    ".repeat(indent_level);

    if let Some(first_child) = children.next() {
        if first_child.kind() == "expression_statement" {
            let mut expr_cursor = first_child.walk();
            for string_node in first_child.named_children(&mut expr_cursor) {
                if string_node.kind() == "string" {
                    // This is the docstring
                    let docstring_text = string_node.utf8_text(source_code.as_bytes()).unwrap();
                    // Strip the quotes and get the quote type
                    let (stripped_docstring, quote_type) = strip_quotes(docstring_text);
                    // Indent the docstring with the correct indentation and re-add quotes
                    let indented_docstring =
                        indent_docstring(stripped_docstring, &indent, quote_type);
                    return indented_docstring;
                }
            }
        }
    }
    String::new()
}

/// Strips quotes around a string literal and returns both the stripped text and the type of quotes used.
fn strip_quotes(s: &str) -> (&str, &str) {
    let s = s.trim();
    if (s.starts_with("\"\"\"") && s.ends_with("\"\"\""))
        || (s.starts_with("'''") && s.ends_with("'''"))
    {
        (&s[3..s.len() - 3], &s[..3]) // Return the inner content and the quote type
    } else if (s.starts_with("\"") && s.ends_with("\"")) || (s.starts_with("'") && s.ends_with("'"))
    {
        (&s[1..s.len() - 1], &s[..1])
    } else {
        (s, "")
    }
}

/// Re-indents a docstring by removing common leading whitespace and re-adding quotes if necessary.
fn indent_docstring(docstring: &str, indent: &str, quote_type: &str) -> String {
    let dedented_docstring = dedent(docstring);

    // Indent each line with the desired indentation
    let indented_docstring = dedented_docstring
        .lines()
        .map(|line| format!("{}{}", indent, line))
        .collect::<Vec<String>>()
        .join("\n");

    // Re-add the quotes with proper indentation
    if quote_type.is_empty() {
        indented_docstring
    } else {
        // For single-line docstrings, keep it in one line
        if !dedented_docstring.contains('\n') {
            format!(
                "{}{}{}{}",
                indent,
                quote_type,
                dedented_docstring.trim(),
                quote_type
            )
        } else {
            // For multi-line docstrings, place quotes on separate lines
            format!(
                "{}{}\n{}\n{}{}",
                indent, quote_type, indented_docstring, indent, quote_type
            )
        }
    }
}

/// Dedents a string by removing leading whitespace from each line and
/// optionally removing empty lines at the start.
fn dedent(s: &str) -> String {
    let lines: Vec<&str> = s.lines().collect();
    // remove first index if it is empty or whitespace-only
    let lines = if lines.first().map_or(true, |line| line.trim().is_empty()) {
        &lines[1..]
    } else {
        &lines
    };

    // remove all whitespace from the beginning of each line
    let string = lines
        .iter()
        .map(|line| line.trim_start())
        .collect::<Vec<&str>>()
        .join("\n");

    string
}

/// Attempts to read an .ipynb file from disk and parse it as JSON.
/// This helper function is used inside the main code to process Jupyter notebooks if needed.
pub fn maybe_read_notebook(file_path: &str) -> Option<JsonValue> {
    if let Ok(notebook_contents) = fs::read_to_string(file_path) {
        if let Ok(notebook_json) = serde_json::from_str::<JsonValue>(&notebook_contents) {
            return Some(notebook_json);
        }
    }
    None
}
