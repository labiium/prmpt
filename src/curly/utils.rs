//! Contains utility functions for handling ignore patterns, directory structure,
//! and file handling for building prompts.

use glob::Pattern;
use std::{fs, io, path::Path};
use walkdir::WalkDir;

/// Reads a `.gitignore` file from the repository path, returning the list of patterns.
/// Returns an error if the file cannot be read, or an empty list if the file does not exist.
pub fn get_gitignore_patterns(repo_path: &Path) -> Result<Vec<Pattern>, io::Error> {
    let gitignore_path = repo_path.join(".gitignore");
    let mut patterns = Vec::new();
    if gitignore_path.exists() {
        let contents = fs::read_to_string(gitignore_path)?;
        for line in contents.lines() {
            let trimmed_line = line.trim();
            if !trimmed_line.is_empty() && !trimmed_line.starts_with('#') {
                patterns.push(Pattern::new(trimmed_line).unwrap());
            }
        }
    }
    Ok(patterns)
}

/// Returns a vector of default patterns to ignore based on the given language name.
pub fn get_default_ignore_patterns(language: &str) -> Vec<Pattern> {
    match language.to_lowercase().as_str() {
        "rust" => vec![
            Pattern::new("target").unwrap(),
            Pattern::new("*.rs.bk").unwrap(),
            Pattern::new("Cargo.lock").unwrap(),
        ],
        "python" => vec![
            Pattern::new("*.pyc").unwrap(),
            Pattern::new("__pycache__").unwrap(),
            Pattern::new(".venv").unwrap(),
            Pattern::new("venv").unwrap(),
        ],
        "javascript" | "typescript" => vec![
            Pattern::new("node_modules").unwrap(),
            Pattern::new("*.min.js").unwrap(),
            Pattern::new("dist").unwrap(),
        ],
        "java" => vec![
            Pattern::new("*.class").unwrap(),
            Pattern::new("*.jar").unwrap(),
            Pattern::new("target").unwrap(),
            Pattern::new(".idea").unwrap(),
        ],
        "c++" => vec![
            Pattern::new("*.o").unwrap(),
            Pattern::new("*.obj").unwrap(),
            Pattern::new("*.exe").unwrap(),
            Pattern::new("build").unwrap(),
        ],
        "go" => vec![
            Pattern::new("*.out").unwrap(),
            Pattern::new("*.test").unwrap(),
            Pattern::new("vendor").unwrap(),
        ],
        "php" => vec![
            Pattern::new("*.log").unwrap(),
            Pattern::new("vendor").unwrap(),
            Pattern::new("composer.lock").unwrap(),
        ],
        _ => vec![],
    }
}

/// Determines if a path should be ignored based on a set of ignore patterns.
/// Handles glob patterns specially to respect directory boundaries.
pub fn should_ignore(path: &Path, base_path: &Path, ignore_patterns: &[Pattern]) -> bool {
    // Convert to a relative path for pattern matching
    let relative_path = match path.strip_prefix(base_path) {
        Ok(p) => p.to_string_lossy(),
        Err(_) => return false,
    };
    
    let relative_path_str = relative_path.to_string();
    
    for pattern in ignore_patterns {
        let pattern_str = pattern.as_str();
        
        // Handle directory-specific patterns (e.g., src/*.rs)
        if pattern_str.contains('/') && pattern_str.contains('*') {
            // Split at the last slash to separate directory part from file pattern
            if let Some(last_slash_pos) = pattern_str.rfind('/') {
                let dir_part = &pattern_str[..=last_slash_pos]; // Include the slash
                let file_pattern = &pattern_str[last_slash_pos + 1..];
                
                // If the pattern ends with a slash (directory only pattern)
                if file_pattern.is_empty() {
                    if relative_path_str.starts_with(dir_part) {
                        return true;
                    }
                    continue;
                }
                
                // Check if path starts with the directory part
                if relative_path_str.starts_with(dir_part) {
                    // Get the remaining part of the path after the directory
                    let remaining_path = &relative_path_str[dir_part.len()..];
                    
                    // Ensure there are no additional slashes (not in subdirectories)
                    if !remaining_path.contains('/') {
                        // Match the file pattern against the remaining path
                        if let Ok(file_glob) = Pattern::new(file_pattern) {
                            if file_glob.matches(remaining_path) {
                                return true;
                            }
                        }
                    }
                    continue;
                }
            }
        }
        
        // Standard full pattern matching
        if pattern.matches(&relative_path) {
            return true;
        }
    }
    
    false
}

/// Recursively builds a textual structure visualization for a directory.
/// This is used to output the tree-like structure seen in the generated prompt.
pub fn process_directory_structure(
    dir: &Path,
    output: &std::sync::Arc<std::sync::Mutex<String>>,
    depth: usize,
    ignore_patterns: &[Pattern],
    prefix: &str,
    base_path: &Path,
) {
    let entries: Vec<_> = WalkDir::new(dir)
        .min_depth(1)
        .max_depth(1)
        .into_iter()
        .filter_map(|e| e.ok())
        .filter(|e| !should_ignore(e.path(), base_path, ignore_patterns))
        .collect();

    for (i, entry) in entries.iter().enumerate() {
        let path = entry.path();
        let is_last = i == entries.len() - 1;

        if path.is_dir() {
            let dir_name = path.file_name().unwrap().to_string_lossy();
            {
                let mut output = output.lock().unwrap();
                output.push_str(&format!(
                    "{}{}── {}\n",
                    prefix,
                    if is_last { "└" } else { "├" },
                    dir_name
                ));
            }
            let new_prefix = format!("{}{}   ", prefix, if is_last { " " } else { "│" });
            process_directory_structure(
                path,
                output,
                depth + 1,
                ignore_patterns,
                &new_prefix,
                base_path,
            );
        } else if path.is_file() {
            let file_name = path.file_name().unwrap().to_string_lossy();
            let mut output = output.lock().unwrap();
            output.push_str(&format!(
                "{}{}── {}\n",
                prefix,
                if is_last { "└" } else { "├" },
                file_name
            ));
        }
    }
}
