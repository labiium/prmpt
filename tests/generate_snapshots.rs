use curly::{Config, Generator, GenerateOperation}; // Corrected crate name
use insta::assert_yaml_snapshot;
use std::path::PathBuf;

// Helper function to construct path to test_repos
fn get_test_repo_path(repo_name: &str) -> PathBuf {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("tests/snapshot_tests/test_repos");
    path.push(repo_name);
    path
}

#[test]
fn test_sample_project_1_default_snapshot() {
    let project_path = get_test_repo_path("sample_project_1");

    let config = Config {
        path: Some(project_path.to_string_lossy().to_string()),
        output: None, // We don't need to write a file for snapshot testing the output string
        ignore: Some(vec![]), // Default: no additional ignores beyond .gitignore
        delimiter: Some("```".to_string()),
        language: Some("python".to_string()), // Explicitly set for clarity
        docs_comments_only: Some(false),      // Default behavior
        docs_ignore: Some(vec![]),
        use_gitignore: Some(true),           // Test .gitignore processing
        display_outputs: Some(false),
        prompts: None,
    };

    let generator = Generator::default();
    let result = generator.run(&config);

    assert!(result.is_ok(), "Generator run failed: {:?}", result.err());
    let (output_string, errors) = result.unwrap();
    
    // Assert that there are no non-critical errors reported from the run
    // (e.g., files that couldn't be processed but didn't stop the whole operation)
    // Depending on strictness, this might be active or commented out.
    // For now, let's ensure it's empty for this controlled test case.
    assert!(errors.is_empty(), "Generator run reported errors: {:?}", errors);

    // Snapshot the main output string
    // The snapshot name will be derived from the test function name:
    // `generate_snapshots__test_sample_project_1_default_snapshot.snap`
    assert_yaml_snapshot!(output_string);
}

#[test]
fn test_sample_project_1_docs_only_snapshot() {
    let project_path = get_test_repo_path("sample_project_1");

    let config = Config {
        path: Some(project_path.to_string_lossy().to_string()),
        output: None,
        ignore: Some(vec![]),
        delimiter: Some("```".to_string()),
        language: Some("python".to_string()),
        docs_comments_only: Some(true), // Test docs_comments_only feature
        docs_ignore: Some(vec![]),
        use_gitignore: Some(true),
        display_outputs: Some(false),
        prompts: None,
    };

    let generator = Generator::default();
    let result = generator.run(&config);

    assert!(result.is_ok(), "Generator run failed for docs_only: {:?}", result.err());
    let (output_string, errors) = result.unwrap();
    assert!(errors.is_empty(), "Generator run for docs_only reported errors: {:?}", errors);

    // Snapshot for the docs_comments_only output
    // Snapshot name: `generate_snapshots__test_sample_project_1_docs_only_snapshot.snap`
    assert_yaml_snapshot!(output_string);
}

#[test]
fn test_config_and_ignore_snapshot() {
    let project_path_str = get_test_repo_path("config_and_ignore_test")
        .to_string_lossy()
        .to_string();

    // Change current directory to the test project path so curly.yaml is found
    let original_dir = std::env::current_dir().unwrap();
    std::env::set_current_dir(&project_path_str).unwrap();

    // Load config from curly.yaml in the test project directory
    let configs = curly::load_config().expect("Failed to load curly.yaml for config_and_ignore_test");
    let config = configs.get(curly::DEFAULT_CONFIG_KEY) // Using DEFAULT_CONFIG_KEY as per new format
        .expect("Config not found under default key for config_and_ignore_test")
        .clone(); // Clone to own it

    // Restore original directory
    std::env::set_current_dir(original_dir).unwrap();

    // Ensure the path in the loaded config is updated to be absolute, or relative to the original dir
    // For simplicity, let's assume load_config doesn't populate `path` if it's not in YAML,
    // or if it does, it's relative. We need to ensure the Generator runs in the context of project_path_str.
    // The easiest way is to set it explicitly after loading.
    let mut effective_config = config;
    effective_config.path = Some(project_path_str.clone());
    // Output is set in curly.yaml, but for snapshot string testing, it's not strictly needed.
    // Let's clear it to ensure we are testing the string output, not file writing.
    effective_config.output = None;


    let generator = Generator::default();
    // Pass a reference to the config
    let result = generator.run(&effective_config);

    assert!(result.is_ok(), "Generator run failed: {:?}", result.err());
    let (output_string, errors) = result.unwrap();

    assert!(errors.is_empty(), "Generator run reported errors: {:?}", errors);

    // Snapshot name will be `generate_snapshots__test_config_and_ignore_snapshot.snap`
    assert_yaml_snapshot!(output_string);
}

#[test]
fn test_config_and_ignore_false_snapshot() {
    let project_path_str = get_test_repo_path("config_and_ignore_false_test")
        .to_string_lossy()
        .to_string();

    // Change current directory to the test project path so curly.yaml is found
    let original_dir = std::env::current_dir().unwrap();
    std::env::set_current_dir(&project_path_str).unwrap();

    // Load config from curly.yaml in the test project directory
    let configs = curly::load_config().expect("Failed to load curly.yaml for config_and_ignore_false_test");
    let config = configs.get(curly::DEFAULT_CONFIG_KEY)
        .expect("Config not found under default key for config_and_ignore_false_test")
        .clone();

    // Restore original directory
    std::env::set_current_dir(original_dir).unwrap();

    let mut effective_config = config;
    effective_config.path = Some(project_path_str.clone());
    effective_config.output = None; // For snapshot string testing

    let generator = Generator::default();
    let result = generator.run(&effective_config);

    assert!(result.is_ok(), "Generator run failed: {:?}", result.err());
    let (output_string, errors) = result.unwrap();

    assert!(errors.is_empty(), "Generator run reported errors: {:?}", errors);

    // Snapshot name will be `generate_snapshots__test_config_and_ignore_false_snapshot.snap`
    assert_yaml_snapshot!(output_string);
}

#[test]
fn output_file_ignorance_snapshot() {
    let test_dir = get_test_repo_path("output_file_ignorance_test");
    let config = Config {
        path: Some(test_dir.to_str().unwrap().to_string()),
        output: Some("test_run_output.out".to_string()), // This file should also be ignored.
        ignore: None, // No specific additional ignores for this test from config
        delimiter: Some("```".to_string()),
        language: Some("rust".to_string()), // Or generic, doesn't matter much for this test
        prompts: None,
        docs_comments_only: Some(false),
        docs_ignore: None,
        use_gitignore: Some(false), // Focus on *.out and curly.yaml ignores
        display_outputs: Some(false),
    };

    let generator = Generator;
    match generator.run(&config) {
        Ok((output, _errors)) => {
            // Normalize paths in the output for consistent snapshots
            // It's crucial to replace the absolute path part with a placeholder.
            // The `get_test_repo_path` gives an absolute path. We need to strip this.
            let repo_root_string = test_dir.to_string_lossy().to_string();
            let normalized_output = output.replace(&repo_root_string, "TEST_REPO_ROOT");
            
            // Further normalization: replace backslashes on Windows if any occur in paths
            let normalized_output = normalized_output.replace("\\", "/");

            // Snapshot name will be: output_file_ignorance_snapshot
            insta::assert_snapshot!("output_file_ignorance_snapshot", normalized_output);
        }
        Err(e) => {
            panic!("Failed to run generator for output_file_ignorance_test: {:?}", e);
        }
    }
}
