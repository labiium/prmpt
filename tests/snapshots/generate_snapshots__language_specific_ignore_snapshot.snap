---
source: tests/generate_snapshots.rs
expression: normalized_output
---
python_ignore_test
└── main.py

```main.py
def hello():
    print("hello")

```
